pkg/lib/bad/error.go

```diff
@@ -0,0 +1,184 @@
+// Package bad helps you deal with errors.
+package bad
+
+import (
+	"strings"
+)
+
+// ErrorSubject identifies where the error arose according to the component that
+// emitted it.
+type ErrorSubject string
+
+const (
+	// The error arose for an unspecified reason.
+	ErrSubjectNone ErrorSubject = ""
+	// The error arose because user input did not conform to the required
+	// specification. The user will need to modify their request before
+	// submitting it again.
+	ErrSubjectInput ErrorSubject = "input"
+	// The error arose because of an internal logic problem with the component
+	// that created the error, normally signaling a programming issue. The
+	// system developer will need to modify their code to meet this request.
+	ErrSubjectInternal ErrorSubject = "internal"
+	// The error arose because some other service or component that the
+	// component that created the error depends on also returned an error. The
+	// error may be transient, especially if this error is a leaf, or may
+	// require operator intervention (e.g. out of disk space).
+	ErrSubjectDependency ErrorSubject = "dependency"
+)
+
+// Error represents a node in a tree of errors. Error (and thus the whole tree)
+// is JSON marshallable so that a tree of Errors can be sent across HTTP.
+type Error struct {
```

Comment:
Can this type be used with the builtin methods `errors.Is` and `errors.As`? If so how?

Reply:
It is a valid `error` so will work fine with `Is` and `As`. 

Do you mean "will `errors.Is(typeOfWrappedError)` return true if I previously called `bad.ToError(typeOfWrappedError)"? If so, no, that would require this class to implement its own Is or As methods. It's not how I envisage this being used.